scripts/test_solidity_lsp.py

```diff
@@ -0,0 +1,874 @@
+            'params': params
+        }
+        json_string = json.dumps(obj=message)
+        rpc_message = f"Content-Length: {len(json_string)}\r\n\r\n{json_string}"
+        self.trace(f'send_message ({method_name})', json.dumps(message, indent=4, sort_keys=True))
+        self.process.stdin.write(rpc_message.encode("utf-8"))
+        self.process.stdin.flush()
+
+    def call_method(self, method_name: str, params: Optional[dict]) -> Any:
+        self.send_message(method_name, params)
+        return self.receive_message()
+
+    def send_notification(self, name: str, params: Optional[dict] = None) -> None:
+        self.send_message(name, params)
+
+# }}}
+
+SGR_RESET = '\033[m'
+SGR_TRACE = '\033[1;36m'
+SGR_NOTICE = '\033[1;35m'
+SGR_TEST_BEGIN = '\033[1;33m'
+SGR_ASSERT_BEGIN = '\033[1;34m'
+SGR_STATUS_OKAY = '\033[1;32m'
+SGR_STATUS_FAIL = '\033[1;31m'
+
+class ExpectationFailed(Exception):
+    def __init__(self, actual, expected):
+        self.actual = actual
+        self.expected = expected
+        diff = DeepDiff(actual, expected)
+        super().__init__(
+            f"Expectation failed.\n\tExpected {expected}\n\tbut got {actual}.\n\t{diff}"
+        )
+
+def create_cli_parser() -> argparse.ArgumentParser:
+    parser = argparse.ArgumentParser(description="Solidity LSP Test suite")
+    parser.set_defaults(trace_io=False)
+    parser.add_argument(
+        "-T, --trace-io",
+        dest="trace_io",
+        action="store_true",
+        help="Be more verbose by also printing assertions."
+    )
+    parser.set_defaults(print_assertions=False)
+    parser.add_argument(
+        "-v, --print-assertions",
+        dest="print_assertions",
+        action="store_true",
+        help="Be more verbose by also printing assertions."
+    )
+    parser.add_argument(
+        "-t, --test-pattern",
+        dest="test_pattern",
+        type=str,
+        default="*",
+        help="Filters all available tests by matching against this test pattern (using globbing)",
+        nargs="?"
+    )
+    parser.add_argument(
+        "solc_path",
+        type=str,
+        default="solc",
+        help="Path to solc binary to test against",
+        nargs="?"
+    )
+    parser.add_argument(
+        "project_root_dir",
+        type=str,
+        default=f"{os.path.dirname(os.path.realpath(__file__))}/..",
+        help="Path to Solidity project's root directory (must be fully qualified).",
+        nargs="?"
+    )
+    return parser
+
+class Counter:
+    total: int = 0
+    passed: int = 0
+    failed: int = 0
+
+class SolidityLSPTestSuite: # {{{
+    test_counter = Counter()
+    assertion_counter = Counter()
+    print_assertions: bool = False
+    trace_io: bool = False
+    test_pattern: str
+
+    def __init__(self):
+        colorama.init()
+        args = create_cli_parser().parse_args()
+        self.solc_path = args.solc_path
+        self.project_root_dir = os.path.realpath(args.project_root_dir) + "/test/libsolidity/lsp"
```

Comment:
The thing I mentioned in https://github.com/ethereum/solidity/pull/11350#discussion_r760500062 is still relevant here:

> Seems a bit misleading to call the lsp dir "project dir". If I saw it, I'd assume it it was the main project dir.

```suggestion
        self.lsp_test_dir = os.path.realpath(args.project_root_dir) + "/test/libsolidity/lsp"
```

Reply:
this is all about lsp, i'm not sure `lsp_test_dir` is actually anything better here, because the reason it is named `project_root_dir` is, because that's what is passed as project root dir to the LSP server.

Reply:
@christianparpart clarified that "project" here means the client's project that sources belong to, and not Solidity as a project. This was not obvious to me at all because I deal with our scripts a lot and the first thing that comes to mind when I see "project root" is our repo root. I still think it should be renamed because it's too ambiguous. How about this?
```suggestion
        self.lsp_project_dir = os.path.realpath(args.project_root_dir) + "/test/libsolidity/lsp"
```

Reply:
`test_workspace_dir`?

Reply:
If workspace is synonymous with project for LSP (I'm not sure about the terminology) then yeah, that's much better.